src/java.base/share/classes/jdk/internal/classfile/impl/ClassfileImpl.java

```diff
@@ -0,0 +1,138 @@
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package jdk.internal.classfile.impl;
+
+import java.util.List;
+import java.util.function.Function;
+import java.util.function.Consumer;
+
+import jdk.internal.classfile.AttributeMapper;
+import jdk.internal.classfile.Classfile;
+import jdk.internal.classfile.Classfile.*;
+import jdk.internal.classfile.ClassBuilder;
+import jdk.internal.classfile.ClassHierarchyResolver;
+import jdk.internal.classfile.ClassModel;
+import jdk.internal.classfile.ClassTransform;
+import jdk.internal.classfile.constantpool.ClassEntry;
+import jdk.internal.classfile.constantpool.ConstantPoolBuilder;
+import jdk.internal.classfile.constantpool.Utf8Entry;
+
+public record ClassfileImpl(StackMapsOption stackMapsOption,
+                            DebugElementsOption debugElementsOption,
+                            LineNumbersOption lineNumbersOption,
+                            UnknownAttributesOption unknownAttributesOption,
+                            ConstantPoolSharingOption constantPoolSharingOption,
+                            ShortJumpsOption shortJumpsOption,
+                            DeadCodeOption deadCodeOption,
+                            DeadLabelsOption deadLabelsOption,
+                            ClassHierarchyResolverOption classHierarchyResolverOption,
+                            AttributeMapperOption attributeMapperOption) implements Classfile {
+
+    public static final ClassfileImpl DEFAULT_CONTEXT = new ClassfileImpl(
+            StackMapsOption.STACK_MAPS_WHEN_REQUIRED,
+            DebugElementsOption.PASS_DEBUG,
+            LineNumbersOption.PASS_LINE_NUMBERS,
+            UnknownAttributesOption.PASS_UNKNOWN_ATTRIBUTES,
+            ConstantPoolSharingOption.SHARED_POOL,
+            ShortJumpsOption.FIX_SHORT_JUMPS,
+            DeadCodeOption.PATCH_DEAD_CODE,
+            DeadLabelsOption.FAIL_ON_DEAD_LABELS,
+            new ClassHierarchyResolverOptionImpl(ClassHierarchyResolver.defaultResolver()),
```

Comment:
Don't users of the default context still share a non-thread-safe-cached version of the default resolver? We might still need to convert DEFAULT_CONTEXT field to a static defaultContext() method that returns a `new ClassfileImpl(...);` (the attribute mapper option can be shared).

Reply:
Please disregard this comment: Now the distinction of cache of resolvers is done in `StackMapGenerator`, meaning each stack map generation task has its own cache. Didn't see it early on.